File: Ejercicios Semana 6/ejercicio2.py
class Empleado:
    def __init__(self, nombre, edad, salario):
        self.nombre = nombre
        self.edad = edad
        self.salario = salario

    def describir_rol(self):
        return f"Soy un empleado de la empresa. Mi nombre es {self.nombre}, tengo {self.edad} años y mi salario es ${self.salario}"

class Gerente(Empleado):
    def __init__(self, nombre, edad, salario, departamento):
        super().__init__(nombre, edad, salario)
        self.departamento = departamento

    def describir_rol(self):
        return f"Soy un gerente de la empresa. Mi nombre es {self.nombre}, tengo {self.edad} años, mi salario es ${self.salario} y estoy a cargo del departamento {self.departamento}"

class Ingeniero(Empleado):
    def __init__(self, nombre, edad, salario, especialidad):
        super().__init__(nombre, edad, salario)
        self.especialidad = especialidad

    def describir_rol(self):
        return f"Soy un ingeniero de la empresa. Mi nombre es {self.nombre}, tengo {self.edad} años, mi salario es ${self.salario} y mi especialidad es {self.especialidad}"

class Asistente(Empleado):
    def __init__(self, nombre, edad, salario, supervisor):
        super().__init__(nombre, edad, salario)
        self.supervisor = supervisor

    def describir_rol(self):
        return f"Soy un asistente de la empresa. Mi nombre es {self.nombre}, tengo {self.edad} años, mi salario es ${self.salario} y mi supervisor es {self.supervisor}"

def interactuar_con_sistema():
    empleados = []
    print("Bienvenido al Sistema de Empleados")
    while True:
        print("\nOpciones:")
        print("1. Crear un empleado")
        print("2. Describir el rol de un empleado")
        print("3. Salir")
        opcion = input("Seleccione una opción: ")

        if opcion == "1":
            nombre = str(input("Nombre del empleado: "))
            edad = int(input("Edad del empleado: "))
            salario = float(input("Salario del empleado: "))
            rol = str(input("Rol del empleado (Gerente/Ingeniero/Asistente): "))
            if rol.lower() == "gerente":
                departamento = int(input("Número del departamento del gerente: "))
                empleado = Gerente(nombre, edad, salario, departamento)
            elif rol.lower() == "ingeniero":
                especialidad = input("Especialidad del ingeniero: ")
                empleado = Ingeniero(nombre, edad, salario, especialidad)
            elif rol.lower() == "asistente":
                supervisor = input("Supervisor del asistente: ")
                empleado = Asistente(nombre, edad, salario, supervisor)
            else:
                print("Rol no válido. Por favor, elija entre Gerente, Ingeniero o Asistente.")
                continue
            empleados.append(empleado)  
            print(f"Empleado creado: {empleado.describir_rol()}")

        elif opcion == "2":
            nombre = input("Nombre del empleado: ")
            encontrado = False
            for empleado in empleados:
                if empleado.nombre == nombre:
                    print(empleado.describir_rol())
                    encontrado = True
                    break
            if not encontrado:
                print(f"No se encontró ningún empleado con el nombre {nombre}")

        elif opcion == "3":
            print("Saliendo del Sistema de Empleados de una Empresa.")
            break

        else:
            print("Opción no válida. Por favor, seleccione una opción válida.")

interactuar_con_sistema()

File: Ejercicios Semana 6/ejercicio2.cpp
#include <iostream>
#include <vector>
#include <string>

class Empleado {
public:
    Empleado(std::string nombre, int edad, float salario)
        : nombre(nombre), edad(edad), salario(salario) {}

    virtual std::string describirRol() {
        return "Soy un empleado de la empresa. Mi nombre es " + nombre +
               ", tengo " + std::to_string(edad) + " años y mi salario es $" +
               std::to_string(salario);
    }

protected:
    std::string nombre;
    int edad;
    float salario;
};

class Gerente : public Empleado {
public:
    Gerente(std::string nombre, int edad, float salario, std::string departamento)
        : Empleado(nombre, edad, salario), departamento(departamento) {}

    std::string describirRol() override {
        return "Soy un gerente de la empresa. Mi nombre es " + nombre +
               ", tengo " + std::to_string(edad) + " años, mi salario es $" +
               std::to_string(salario) + " y estoy a cargo del departamento " +
               departamento;
    }

private:
    std::string departamento;
};

class Ingeniero : public Empleado {
public:
    Ingeniero(std::string nombre, int edad, float salario, std::string especialidad)
        : Empleado(nombre, edad, salario), especialidad(especialidad) {}

    std::string describirRol() override {
        return "Soy un ingeniero de la empresa. Mi nombre es " + nombre +
               ", tengo " + std::to_string(edad) + " años, mi salario es $" +
               std::to_string(salario) + " y mi especialidad es " + especialidad;
    }

private:
    std::string especialidad;
};

class Asistente : public Empleado {
public:
    Asistente(std::string nombre, int edad, float salario, std::string supervisor)
        : Empleado(nombre, edad, salario), supervisor(supervisor) {}

    std::string describirRol() override {
        return "Soy un asistente de la empresa. Mi nombre es " + nombre +
               ", tengo " + std::to_string(edad) + " años, mi salario es $" +
               std::to_string(salario) + " y mi supervisor es " + supervisor;
    }

private:
    std::string supervisor;
};

void interactuarConSistema() {
    std::vector<Empleado*> empleados;
    std::cout << "Bienvenido al Sistema de Empleados" << std::endl;
    while (true) {
        std::cout << "\nOpciones:" << std::endl;
        std::cout << "1. Crear un empleado" << std::endl;
        std::cout << "2. Describir el rol de un empleado" << std::endl;
        std::cout << "3. Salir" << std::endl;
        std::cout << "Seleccione una opcion: ";
        std::string opcion;
        std::cin >> opcion;

        if (opcion == "1") {
            std::string nombre;
            int edad;
            float salario;
            std::string rol;

            std::cout << "Nombre del empleado: ";
            std::cin >> nombre;
            std::cout << "Edad del empleado: ";
            std::cin >> edad;
            std::cout << "Salario del empleado: ";
            std::cin >> salario;
            std::cout << "Rol del empleado (Gerente/Ingeniero/Asistente): ";
            std::cin >> rol;

            if (rol == "Gerente" || rol == "gerente") {
                std::string departamento;
                std::cout << "Numero del departamento del gerente: ";
                std::cin >> departamento;
                empleados.push_back(new Gerente(nombre, edad, salario, departamento));
            } else if (rol == "Ingeniero" || rol == "ingeniero") {
                std::string especialidad;
                std::cout << "Especialidad del ingeniero: ";
                std::cin >> especialidad;
                empleados.push_back(new Ingeniero(nombre, edad, salario, especialidad));
            } else if (rol == "Asistente" || rol == "asistente") {
                std::string supervisor;
                std::cout << "Supervisor del asistente: ";
                std::cin >> supervisor;
                empleados.push_back(new Asistente(nombre, edad, salario, supervisor));
            } else {
                std::cout << "Rol no valido. Por favor, elija entre Gerente, Ingeniero o Asistente." << std::endl;
            }

            std::cout << "Empleado creado." << std::endl;
        } else if (opcion == "2") {
            std::string nombre;
            std::cout << "Nombre del empleado: ";
            std::cin >> nombre;
            bool encontrado = false;
            for (Empleado* empleado : empleados) {
                if (empleado->describirRol().find(nombre) != std::string::npos) {
                    std::cout << empleado->describirRol() << std::endl;
                    encontrado = true;
                    break;
                }
            }
            if (!encontrado) {
                std::cout << "No se encontro ningun empleado con el nombre " << nombre << std::endl;
            }
        } else if (opcion == "3") {
            std::cout << "Saliendo del Sistema de Empleados de una Empresa." << std::endl;
            break;
        } else {
            std::cout << "Opcion no valida. Por favor, seleccione una opcion valida." << std::endl;
        }
    }
}

int main() {
    interactuarConSistema();
    return 0;
}
